src/widgets/helper/SearchPopup.cpp

```diff
@@ -113,4 +129,78 @@ void SearchPopup::initLayout()
     }
 }
 
+std::vector<MessagePredicatePtr> SearchPopup::parsePredicates(
+    const QString &input)

```

Comment:
The functions here are repetitive: The text gets split and iterated in 3 different functions and there seem to be multiple ways to find key:value tags.

I suggest something like this:
```c++
std::vector<MessagePredicatePtr> SearchPopup::parsePredicates(
    const QString &input)
{
    static QRegularExpression predicateRegex(R"(^(\w+):(\w+)$)");

    auto words = input.split(' ', QString::SkipEmptyParts);
    // turn shared into unique ptr later!
    auto preds = std::vector<std::shared_ptr<MessagePredicate>>();
    auto authors = QStringList();

    for (auto it = words.begin(); it != words.end();)
    {
        if (auto match = predicateRegex.match(*it); match.hasMatch())
        {
            QString name = match.captured(1);
            QString value = match.captured(2);

            bool remove = true;

            // match predicates
            if (name == "from")
            {
                authors.append(value);
            }
            else if (name == "has" && value == "link")
            {
                preds.push_back(std::make_unique<LinkPredicate>());
            }
            else
            {
                remove = false;
            }

            // remove or advance
            it = remove ? words.erase(it) : ++it;
        }
    }

    if (!authors.empty())
        preds.push_back(std::make_unique<AuthorPredicate>(authors));

    if (!words.empty())
        preds.push_back(std::make_unique<SubstringPredicate>(words.join(" ")));

    return preds;
}

```

Reply:
Resolved in 6e9dbc1e.